### PR TITLE
Feat rstar2: add rstar2 support

### PR DESCRIPTION
**Pull request by nufukim**
rstar2-agent RL training recipe

### Description

fix issue [support rstar2-agent](https://github.com/RLinf/RLinf/issues/78)
I implemented the simple rstar2-agent algorithm.
- agentic rollout with the Python execution environment:
    A simple data distribution method is used, without involving load balancing. 
    Use sglang as rollout backend.

- fused multi- math verification tools during reward stage:
    It can be called correctly now, but it has not been tested with actual data.

- add rollout trace data augmentation stage after rollout & reward stage
    The sampling logic has been implemented, but there is a conflict with the batch size processing in the mother framework. We are still discussing how to solve it

- align training hyperparameters
    TODO:...

### Motivation and Context
link: [paper](https://www.arxiv.org/pdf/2508.20722).

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/utils/resharding/utils.py [hunk 1]
```diff
@@ -9,7 +9,10 @@
 # Unless required by applicable law or agreed to in writing, software
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
+# See the License for the specific lan
+# 
+#
```
Comment (qurakchin) [inline rlinf/utils/resharding/utils.py]: license comment is broken

File: rlinf/workers/actor/megatron_actor_worker.py [hunk 2]
```diff
@@ -682,14 +694,28 @@ def run_training(self, input_channel: Channel):
         # Get all batches for this DP
         batches = []
         recv_batch_size = 0
-        while recv_batch_size < self.total_batch_size_per_dp:
+        while recv_batch_size < self.after_sample_total_batch_size_per_dp:
```
Comment (qurakchin) [inline rlinf/workers/actor/megatron_actor_worker.py]: The variable use original name 'total_batch_size_per_dp', and add a comment in __init__ to describe that the variable may change between every iter.

File: rlinf/workers/actor/megatron_actor_worker.py [hunk 3]
```diff
@@ -146,9 +147,17 @@ def __init__(
         # Reward configurations
         if not self.cfg.reward.use_reward_model:
             assert self.cfg.reward.reward_type == "math", "only support math"
-            self.reward_fn = math_verify_call
-
-        # Rollout configurations
+            # 绑定本地reward函数：使用vendored compute_score
```
Comment (qurakchin) [inline rlinf/workers/actor/megatron_actor_worker.py]: pr #128 will use reward worker. rebase to fit this feature when #128 is merged.
delete non-english comments.

File: rlinf/workers/rollout/utils.py [hunk 4]
```diff
@@ -531,7 +531,7 @@ def _get_actor_rank_to_rollout_rank(
         return (corresponding_rollout_dp_rank, corresponding_rollout_tp_rank)
 
 
-SUPPORTED_LLM_ROLLOUT_BACKENDS = ["vllm", "sglang"]
+SUPPORTED_LLM_ROLLOUT_BACKENDS = ["vllm", "sglang", "agentloop_sglang"]
```
Comment (qurakchin) [inline rlinf/workers/rollout/utils.py:540]: sglang worker is updated and now these change is not need. rebase to main to use latest impl.

File: rlinf/utils/distributed.py [hunk 5]
```diff
@@ -33,8 +33,9 @@
 def compute_rollout_metrics(
     rollout_batch, max_prompt_len, response_len, use_critic=False
 ):
+    print(f"rollout_batch: {rollout_batch}")
     device = torch.device(f"cuda:{torch.cuda.current_device()}")
-    advantages = rollout_batch["advantages"].to(device=device)
+    advantages = rollout_batch["advantages"].to(device=device) if rollout_batch["advantages"] is not None else None
```
Comment (qurakchin) [inline rlinf/utils/distributed.py:42]: Why no advantages? Is rejection sampling caused it?

File: rlinf/utils/data_iter_utils.py [hunk 6]
```diff
@@ -162,6 +162,9 @@ def get_iterator_k_split(
 
         # Split tensor items
         items = list(tensor_items.items())
+        while items[0][1].shape[0] % num_splits != 0:
```
Comment (qurakchin) [inline rlinf/utils/data_iter_utils.py]: add comment for this.

File: rlinf/runners/rstar2_runner.py [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+                batch_timeout_seconds=code_judge_config.get("batch_timeout_seconds", 30.0),
+            )
+        
+        return tools
+
+    # 已移除通道处理器
+
+    # 已移除旧的兼容生成函数
+
+    def _run_agentloop_rollout(self):
```
Comment (qurakchin) [inline rlinf/runners/rstar2_runner.py]: Refactor this into an agentloop worker, and the agentloop worker is an abstruct class and need to be inherent

File: toolkits/rstar2/code-judge-main/.dockerignore [hunk 8]
```diff
@@ -0,0 +1,6 @@
+.git/
```
Comment (qurakchin) [inline toolkits/rstar2/code-judge-main/.dockerignore:1]: Are all files and folders only for rstar2? move this into an sub folder if these are all only for rstar2.

File: rlinf/runners/rstar2_runner.py [hunk 9]
```diff
@@ -0,0 +1,722 @@
...
+        self.cfg = cfg
+        self.component_placement = placement
+        self.is_pipeline = self.component_placement.is_disaggregated
+        self.has_dedicated_inference = inference is not None
+        self.has_dedicated_reward = reward is not None
+        self.has_dedicated_down_sample = down_sample is not None
+
+        # Workers
+        self.actor = actor
+        self.rollout = rollout  # 存储rollout_worker引用
```
Comment (qurakchin) [inline rlinf/runners/rstar2_runner.py:75]: remove non-english comments.

File: rlinf/hybrid_engines/sglang/sglang_0_4_4/sgl_engine.py [hunk 10]
```diff
@@ -104,6 +104,11 @@ def offload_model_weights(self):
             self.tokenizer_manager.offload_model_weights(obj, None)
         )
 
+    async def async_offload_model_weights(self):
```
Comment (qurakchin) [inline rlinf/hybrid_engines/sglang/sglang_0_4_4/sgl_engine.py]: Is this from another branch?

File: rlinf/config.py [hunk 11]
```diff
@@ -175,11 +175,11 @@ def validate_vllm_cfg(cfg):
...
         cfg.rollout_backend = cfg.get("rollout_backend", "sglang")
-        assert cfg.rollout_backend in SUPPORTED_ROLLOUT_BACKENDS, (
-            f"rollout_backend must be one of {SUPPORTED_ROLLOUT_BACKENDS}."
-        )
+        # assert cfg.rollout_backend in SUPPORTED_ROLLOUT_BACKENDS, (
+        #     f"rollout_backend must be one of {SUPPORTED_ROLLOUT_BACKENDS}."
+        # )
         cfg.sglang = validate_sglang_cfg(cfg.sglang)
-        cfg.vllm = validate_vllm_cfg(cfg.vllm)
+        # cfg.vllm = validate_vllm_cfg(cfg.vllm)
```
Comment (qurakchin) [inline rlinf/config.py]: uncomment this and add rollout.vllm in your yaml.

File: rlinf/runners/rstar2_runner.py [hunk 12]
```diff
@@ -466,13 +466,13 @@ def epoch(self):
...
         lengths = batch["length"].tolist()
-        answers = batch["answer"].tolist()
+        answers = batch["answer"]
         image_data = batch["image_data"]
         multi_modal_inputs = batch["multi_modal_inputs"]
         prompts = [ids[-pmp_len:] for ids, pmp_len in zip(prompt_ids, lengths)]
         rollout_dp_size = self.component_placement.rollout_dp_size
 
-        for input_ids, answers in zip(
+        for input_ids, answers, image_data, multi_modal_inputs in zip(
```
Comment (qurakchin) [inline rlinf/runners/rstar2_runner.py]: Does rstar2 uses multi_modal_inputs and image inputs?

Comment (nufukim) [reply rlinf/runners/rstar2_runner.py]: no，these code was from main branch

File: rlinf/config.py (hunk 11 above)
Comment (nufukim) [reply rlinf/config.py]: Done

File: rlinf/hybrid_engines/sglang/sglang_0_4_4/sgl_engine.py (hunk 10 above)
Comment (nufukim) [reply rlinf/hybrid_engines/sglang/sglang_0_4_4/sgl_engine.py]: will be removed

File: rlinf/utils/distributed.py (hunk 5 above)
Comment (nufukim) [reply rlinf/utils/distributed.py:42]: yes

File: rlinf/workers/rollout/utils.py (hunk 4 above)
Comment (nufukim) [reply rlinf/workers/rollout/utils.py:540]: done

File: rlinf/workers/actor/megatron_actor_worker.py (hunk 2 above)
Comment (nufukim) [reply rlinf/workers/actor/megatron_actor_worker.py]: Done

File: rlinf/utils/resharding/utils.py (hunk 1 above)
Comment (nufukim) [reply rlinf/utils/resharding/utils.py]: resolved
